### `---------------Mandatory Information to fill------------`

### Group ID:
### Group Members Name with Student ID:
1. Student 1
2. Student 2
3. Student 3
4. Student 4


`-------------------Write your remarks (if any) that you want should get consider at the time of evaluation---------------`

Remarks: ##Add here

# Background

In the world of online streaming, user satisfaction and engagement are critical metrics for the success of a movie recommendation system. A well-designed recommendation algorithm can significantly enhance user experience by suggesting movies that align with their preferences, leading to higher platform retention and usage. Recommendation systems face the challenge of balancing exploration (discovering new movies) with exploitation (recommending known favourites) to maximize user satisfaction over time.


# Scenario

Imagine a leading online movie streaming platform, TrendMovie Inc., that aims to become the go-to destination for personalized movie recommendations. The platform features a vast collection of movies catering to diverse audiences. TrendMovie Inc. wants to optimize its recommendation strategy to deliver maximum user satisfaction while maintaining a high level of engagement. Each movie recommendation is treated as an interaction with the user, and their feedback is used to refine the recommendation strategy dynamically.


# Objective

Your objective is to design and implement a recommendation system using Multi-Armed Bandit (MAB) algorithms to maximize cumulative user satisfaction. The system should dynamically allocate recommendations by learning user preferences in real-time, striking the right balance between exploration and exploitation.


# Dataset

The dataset contains user ratings for a variety of movies. Key columns in the dataset include:
*   **User ID:** A unique identifier for each user.
*   **Movie ID:** A unique identifier for each.
*   **Rating:** A score provided by the user for a movie (on a scale of 1 to 5).
*   **Timestamp:** The time when the rating was given (optional for this assignment).

***Link for accessing dataset:***
https://drive.google.com/file/d/1gfobhqlVCw8Oo52JCiYpEBGhG5k7cWBr/view?usp=drive_link


# Environment Details

**Arms:** Each movie represents an "arm" in the MAB framework. The probability of a movie being liked by a user is initially unknown and will be estimated based on user feedback during the interactions.
For example:

Arm 1: Movie A

Arm 2: Movie B

Arm 3: Movie C

... and so on, for all movies in the dataset.

**Reward Function:**
The reward function is defined based on user ratings:

***Reward = 1:*** The user rates the movie high star (e.g., 4 or 5 stars).

***Reward = 0:*** The user rates the movie low star (e.g., 1, 2, or 3 stars).


**Assumptions:**

Run simulations for 1000 iterations for each policy


# Requirements and Deliverables:
Implement the Multi-Arm Bandit Problem for the given above scenario for all the below mentioned policy methods.

### Initialize constants

In [37]:
!pip install gym pandas

In [38]:
import random
import numpy as np
import pandas as pd
import gym

# Constants
no_of_iterations = 1000

# Initialize value function and policy

# Load Dataset (0.5M)

In [39]:
# Code for Dataset loading and print dataset statistics
#-----write your code below this line---------


# Data Exploration

import pandas as pd

# Load the dataset
dataset = pd.read_csv("TrendMovie.csv")

# Display basic information about the dataset
print("Dataset Info:")
dataset.info()

print("Total Movie count:")
print(len(dataset['movieId'].unique()))

# Display the first few rows of the dataset
print("\nDataset Preview:")
print(dataset.head())

# Check for missing values
print("\nMissing Values:")
print(dataset.isnull().sum())

# Summary statistics of the ratings
print("\nSummary Statistics:")
print(dataset["rating"].describe())

# Preprocessing

# Convert timestamp to a readable format
dataset["timestamp"] = pd.to_datetime(dataset["timestamp"], unit="s")

# Sort by user and timestamp for consistent processing
dataset = dataset.sort_values(by=["userId", "timestamp"]).reset_index(drop=True)

print("\nProcessed Dataset Preview:")
print(dataset.head())



Dataset Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100836 entries, 0 to 100835
Data columns (total 4 columns):
 #   Column     Non-Null Count   Dtype  
---  ------     --------------   -----  
 0   userId     100836 non-null  int64  
 1   movieId    100836 non-null  int64  
 2   rating     100836 non-null  float64
 3   timestamp  100836 non-null  int64  
dtypes: float64(1), int64(3)
memory usage: 3.1 MB
Total Movie count:
9724

Dataset Preview:
   userId  movieId  rating  timestamp
0       1        1     4.0  964982703
1       1        3     4.0  964981247
2       1        6     4.0  964982224
3       1       47     5.0  964983815
4       1       50     5.0  964982931

Missing Values:
userId       0
movieId      0
rating       0
timestamp    0
dtype: int64

Summary Statistics:
count    100836.000000
mean          3.501557
std           1.042529
min           0.500000
25%           3.000000
50%           3.500000
75%           4.000000
max           5.000000
Name: rating, d

# Design a Movie Environment (0.5M)

In [40]:
import random
import pandas as pd
import gym
import math

# Define the Movie Recommendation Environment
class MovieRecommendationEnv(gym.Env):
    def __init__(self, data):
        """
        Initialize the environment.

        Parameters:
        - data: DataFrame containing user ratings with columns: 'userId', 'movieId', 'rating'.
        """
        self.data = data
        self.movies = data['movieId'].unique()  # Unique movie IDs
        self.state = {movie_id: [] for movie_id in self.movies}  # Tracks user feedback for each movie
        self.reset()  # Initialize the state

    def step(self, action):
        """
        Simulate a user's interaction with a selected movie (arm).

        Parameters:
        - action: The selected movie ID (arm).

        Returns:
        - state: Updated state with rewards for each movie.
        - reward: 1 if liked, 0 otherwise.
        - done: Always False, as this is a continuous environment.
        - debug: None for now (can be extended for additional info).
        """
        # Filter ratings for the selected movie
        movie_ratings = self.data[self.data['movieId'] == action]['rating']
        
        # Use the average rating for the movie
        if not movie_ratings.empty:
            # print(action, 'avg ', movie_ratings.sum(), len(movie_ratings), movie_ratings.mean())
            avg_rating = movie_ratings.mean()
            # avg_rating = movie_ratings.sample().iloc[0]
            reward = 1 if avg_rating >= 4.0 else 0  # High rating gives a reward of 1
        else:
            reward = 0  # Default to 0 if no ratings are available

        self.state[action].append(reward)  # Update state with reward
        done = False
        debug = None
        return self.state, reward, done, debug

    def reset(self):
        """
        Reset the environment state. Each movie starts with an empty feedback list.

        Returns:
        - state: The initial state of the environment.
        """
        self.state = {movie_id: [] for movie_id in self.movies}
        return self.state

    def render(self, mode="ascii"):
        """
        Display statistics of user feedback for all movies.

        Parameters:
        - mode: The rendering mode (default is "ascii").
        """
        returns = {movie_id: sum(rewards) for movie_id, rewards in self.state.items()}
        trials = {movie_id: len(rewards) for movie_id, rewards in self.state.items()}
        print(f'===== Total Trials: {sum(trials.values())} =====')
        for movie_id, total_return in returns.items():
            total_trials = trials[movie_id]
            print(f'Movie {movie_id} | Returns: {total_return}, Trials: {total_trials}')
        print(f'===== Total Returns: {sum(returns.values())} =====')

# Greedy Policy Implementation
def greedy_policy(env, iterations=1000):
    """
    Implement the greedy policy to select movies based on estimated probabilities.

    Parameters:
    - env: The movie recommendation environment.
    - iterations: Number of iterations for the simulation.

    Returns:
    - None
    """
    estimated_values = {movie_id: 0 for movie_id in env.movies}  # Initialize estimated values
    counts = {movie_id: 0 for movie_id in env.movies}  # Track the number of times each movie is selected

    for _ in range(iterations):
        # Select the movie with the highest estimated probability (greedy choice)
        action = max(estimated_values, key=estimated_values.get)

        # Interact with the environment using the selected action
        _, reward, _, _ = env.step(action)

        # Update counts and estimated values
        counts[action] += 1
        estimated_values[action] += (reward - estimated_values[action]) / counts[action]

    # Render final statistics
    env.render()

# Random Policy Implementation
def random_policy(env, iterations=1000):
    """
    Implement the random policy to select movies randomly.

    Parameters:
    - env: The movie recommendation environment.
    - iterations: Number of iterations for the simulation.

    Returns:
    - None
    """
    for _ in range(iterations):
        # Select a random movie
        action = random.choice(env.movies)
        
        print('selected', action)

        # Interact with the environment using the selected action
        _, reward, _, _ = env.step(action)

    # Render final statistics
    env.render()

# Upper Confidence Bound (UCB) Policy Implementation
def ucb_policy(env, iterations=1000, c=2):
    """
    Implement the UCB policy to select movies based on estimated values and confidence bounds.

    Parameters:
    - env: The movie recommendation environment.
    - iterations: Number of iterations for the simulation.
    - c: Exploration parameter for UCB.

    Returns:
    - None
    """
    estimated_values = {movie_id: 0 for movie_id in env.movies}  # Initialize estimated values
    counts = {movie_id: 0 for movie_id in env.movies}  # Track the number of times each movie is selected

    for t in range(1, iterations + 1):
        # Select the movie based on UCB formula
        ucb_values = {}
        for movie_id in env.movies:
            if counts[movie_id] == 0:
                ucb_values[movie_id] = float('inf')  # Select untried movies first
            else:
                ucb_values[movie_id] = estimated_values[movie_id] + c * math.sqrt(math.log(t) / counts[movie_id])

        action = max(ucb_values, key=ucb_values.get)

        # Interact with the environment using the selected action
        _, reward, _, _ = env.step(action)

        # Update counts and estimated values
        counts[action] += 1
        estimated_values[action] += (reward - estimated_values[action]) / counts[action]

    # Render final statistics
    env.render()

# Main Execution
if __name__ == "__main__":
    # Load the dataset
    csv_file_path = "TrendMovie.csv"  # Path to the CSV file
    data = pd.read_csv(csv_file_path)  # Load data from the CSV

    # Initialize the environment with the dataset
    env = MovieRecommendationEnv(data)

    # Run the policies
    print("\nRunning Random Policy:")
    random_policy(env)
    
    # print("Running Greedy Policy:")
    # greedy_policy(env)


    # print("\nRunning UCB Policy:")
    # ucb_policy(env)



Running Random Policy:
selected 250
selected 4454
selected 48596
selected 1936
selected 132422
selected 8952
selected 1083
selected 1214
selected 130842
selected 2789
selected 44204
selected 80363
selected 77800
selected 2559
selected 32116
selected 68522
selected 6962
selected 142456
selected 3567
selected 988
selected 91094
selected 182715
selected 2951
selected 1230
selected 3357
selected 1078
selected 2549
selected 88235
selected 4957
selected 7354
selected 1371
selected 57843
selected 98243
selected 278
selected 31878
selected 6620
selected 99191
selected 133419
selected 5028
selected 2669
selected 58025
selected 4757
selected 30825
selected 6100
selected 51088
selected 8934
selected 110669
selected 5941
selected 6551
selected 71282
selected 2648
selected 84246
selected 4439
selected 131013
selected 6218
selected 1446
selected 2657
selected 65682
selected 100487
selected 125
selected 8601
selected 67734
selected 50800
selected 8959
selected 180263
selected 7991
selected 74696
sel

# Using Random Policy (0.5M)
Implement a random policy for movie recommendations and print each iteration. (Mandatory)

In [41]:
#  run the environment with an agent that is guided by a random policy
#-----write your code below this line---------

# Using Greedy Policy (1M)
Implement a greedy policy that always recommends the movie with the highest estimated reward and print each iteration. (Mandatory)

In [42]:
#  run the environment with an agent that is guided by a greedy policy
#-----write your code below this line---------

# Using Epsilon-Greedy Policy (1.5M)
Implement the epsilon-greedy policy, where with probability ε you explore (recommend a random movie) and with probability (1-ε) you exploit (recommend the best-known movie). Try with ε =0.1, 0.2, 0.5 and print each iteration. What value of ε yields the best performance? (Mandatory)

In [43]:
#  run the environment with an agent that is guided by a epsilon-greedy policy
#-----write your code below this line---------

# Using UCB (1M)
Implement the UCB algorithm for movie recommendations and print each iteration. (Mandatory)

In [44]:
#  run the environment with an agent that is guided by a UCB
#-----write your code below this line---------

# Plot the cumulative rewards for all policies on a single graph to compare their performance. (0.5M)

In [45]:
#-----write your code below this line---------

# Conclusion (0.5M)

Determine which policy performs the best based on cumulative reward. Provide a concise conclusion (250 words) summarizing the decision-making process and the trade-offs between exploration and exploitation.

`----write below this line------`